pkg/controllers/machine/lifecycle/launch.go

```diff
@@ -0,0 +1,147 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lifecycle
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/patrickmn/go-cache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/metrics"
+	"github.com/aws/karpenter-core/pkg/scheduling"
+)
+
+type Launch struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+	cache         *cache.Cache // exists due to eventual consistency on the cache
+}
+
+func (l *Launch) Reconcile(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	if machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).IsTrue() {
+		return reconcile.Result{}, nil
+	}
+
+	var err error
+	var created *v1alpha5.Machine
+
+	if ret, ok := l.cache.Get(client.ObjectKeyFromObject(machine).String()); ok {
+		created = ret.(*v1alpha5.Machine)
+	} else if _, ok := machine.Annotations[v1alpha5.MachineLinkedAnnotationKey]; ok {
+		created, err = l.linkMachine(ctx, machine)
+	} else {
+		created, err = l.launchMachine(ctx, machine)
+	}
+	// Either the machine launch/linking failed or the machine was deleted due to InsufficientCapacity/NotFound
+	if err != nil || created == nil {
+		return reconcile.Result{}, err
+	}
+	l.cache.SetDefault(client.ObjectKeyFromObject(machine).String(), created)
+	PopulateMachineDetails(machine, created)
+	machine.StatusConditions().MarkTrue(v1alpha5.MachineLaunched)
+	return reconcile.Result{}, nil
+}
+
+func (l *Launch) linkMachine(ctx context.Context, machine *v1alpha5.Machine) (*v1alpha5.Machine, error) {
+	created, err := l.cloudProvider.Get(ctx, machine.Annotations[v1alpha5.MachineLinkedAnnotationKey])
+	if err != nil {
+		if cloudprovider.IsMachineNotFoundError(err) {
```

Comment:
nit on nested ifs: 
Can we do short circuiting here and do
```suggestion
		if !cloudprovider.IsMachineNotFoundError(err) {
```

then bring up lines 81 and 82 up, and then do lines 71-79 after?